{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d045538b-a1f1-49a0-b0b9-e58b2d448774",
   "metadata": {},
   "source": [
    "# Summary report exported as a workbook\n",
    "\n",
    "Author: [José R. Ferrer-Paris](https://github.com/jrfep)\n",
    "\n",
    "This notebook:\n",
    "- Reads information from the database, and\n",
    "- Creates a workbook with:\n",
    "    - Authoring information and instruction\n",
    "    - Summary table for species with links\n",
    "    - Trait codes and descriptions\n",
    "    - Vocabularies\n",
    "    - List of references\n",
    "\n",
    "The outputs of this notebook are available as a dataset record at:\n",
    "\n",
    "> Ferrer-Paris, José R.; Keith, D A (2024). Fire Ecology Traits for Plants: Database exports. figshare. Dataset. https://doi.org/10.6084/m9.figshare.24125088.v1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "074c5468-519c-4ea5-a46c-f5e09d57c8a7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Setup\n",
    "\n",
    "These sections include basic set up for the project"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f6f23f7-d797-4dcd-be3a-f90a32774dca",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5626d235-7c71-459c-a529-93658ed61ea8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# work with paths in operating system\n",
    "from pathlib import Path\n",
    "import os\n",
    "import sys\n",
    "# datetime support\n",
    "import datetime\n",
    "\n",
    "# work with xlsx workbooks\n",
    "import openpyxl\n",
    "from openpyxl import Workbook\n",
    "from openpyxl.worksheet.table import Table, TableStyleInfo\n",
    "from openpyxl.styles import Alignment, PatternFill, Border, Font # Side, Alignment, Protection,\n",
    "from openpyxl.formatting import Rule\n",
    "from openpyxl.styles.differential import DifferentialStyle\n",
    "from openpyxl.worksheet.datavalidation import DataValidation\n",
    "\n",
    "from openpyxl.utils.dataframe import dataframe_to_rows\n",
    "from openpyxl.utils import get_column_letter\n",
    "\n",
    "# For database connection\n",
    "from configparser import ConfigParser\n",
    "import psycopg2\n",
    "from psycopg2.extras import DictCursor\n",
    "\n",
    "# Pandas for calculations\n",
    "import pandas as pd\n",
    "# Pyprojroot for easier handling of working directory\n",
    "import pyprojroot"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b004146-868a-4580-afaf-0da9c861f698",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Define paths for input and output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d61e4223-ec48-4d31-9a94-61c002ab8b09",
   "metadata": {},
   "outputs": [],
   "source": [
    "repodir = pyprojroot.find_root(pyprojroot.has_dir(\".git\"))\n",
    "sys.path.append(str(repodir))\n",
    "inputdir = repodir / \"data\" / \"output-report\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84695007-17e1-45d2-a357-51697dcf77f6",
   "metadata": {},
   "source": [
    "### Load own functions\n",
    "Load functions from `lib` folder, we will use a function to read db credentials, one for executing database queries and three functions for extracting data from the reference description string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a40a35a2-f7d3-46d2-a181-c396e8460d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "from lib.parseparams import read_dbparams\n",
    "from lib.firevegdb import dbquery\n",
    "import lib.firevegxport as fvx"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bfaa780-1341-4e8a-96c3-41f9be39993e",
   "metadata": {},
   "source": [
    "## Read information from database"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d828bc4-c9ff-4617-9701-ec07cf72af42",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Database connection parameters\n",
    "Database credentials are stored in a `database.ini` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a0864b79-f8d1-4b3f-bc05-79d78eac3b75",
   "metadata": {},
   "outputs": [],
   "source": [
    "dbparams = read_dbparams(repodir / 'secrets' / 'database.ini', section='aws-lght-sl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c14c50b1-59bc-4e89-8f56-f95c62751cec",
   "metadata": {},
   "source": [
    "### Database queries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e8a814b-31b7-4e21-92a1-bbd5218567d6",
   "metadata": {},
   "source": [
    "The table with trait information can be requested with a simple query:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a265c435-d211-4955-bed3-ef519f45dd1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "qrystr = \"SELECT code,name,description,value_type,life_stage,life_history_process,priority FROM litrev.trait_info ORDER BY code\"\n",
    "trait_info = dbquery(qrystr, dbparams)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3252aa9f-7ae6-4d91-bf58-f7dcaf7421cd",
   "metadata": {
    "tags": []
   },
   "source": [
    "If we want to build a table with all the traits, we need to concatenate the results of several queries. For some of these, we will need a custom array function for [handling empty arrays in postgresql](https://stackoverflow.com/questions/43472482/postgres-array-agg-throws-cannot-accumulate-empty-arrays-for-empty-arrays). So we have to run this on the postgres server side:"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d0b68068-a802-4e5f-89b4-052b5b96e6d1",
   "metadata": {
    "tags": []
   },
   "source": [
    "CREATE AGGREGATE array_accum (anyarray)\n",
    "(\n",
    "    sfunc = array_cat,\n",
    "    stype = anyarray,\n",
    "    initcond = '{}'\n",
    ");  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "258a5329-b00b-4927-95ea-ac18a67c0c64",
   "metadata": {},
   "source": [
    "Now we use a general query string to retrieve selected columns, and we will run a for loop to run the query with different categorical traits, and then use functions from the pandas library to merge the data into a pandas data frame. Here the trait codes are also the names of the tables in the `litrev.` schema:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "962a3bf0-deee-40c9-ab6b-66cca0b0ce6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "qry= \"\"\"\n",
    "SELECT \"currentScientificName\" as spp, \"currentScientificNameCode\" as sppcode,\n",
    "    array_agg(species) as nspp,\n",
    "    array_agg(norm_value::text) as val,array_agg(weight) as w,\n",
    "    array_agg(main_source) as refs,\n",
    "    array_accum(original_sources) as orefs\n",
    "    \n",
    "FROM litrev.{} \n",
    "LEFT JOIN species.caps\n",
    "ON species_code=\"speciesCode_Synonym\"\n",
    "WHERE \"currentScientificName\" is not NULL AND main_source is not NULL AND weight>0\n",
    "GROUP BY spp,sppcode;\n",
    "\"\"\"\n",
    "\n",
    "# test using only ecualipts\n",
    "# WHERE species ilike '%euca%' and \"currentScientificName\" is not NULL AND weight>0\n",
    "\n",
    "for trait in ['surv1','surv4','repr2','rect2','disp1','germ1','germ8']:\n",
    "    res = dbquery(qry.format(trait), dbparams)\n",
    "    \n",
    "    df1 = pd.DataFrame(res)\n",
    "    col1=\"%s.txn\" % trait\n",
    "    col2=\"%s.v\" % trait\n",
    "    col3=\"%s.w\" % trait\n",
    "    col4=\"%s.mref\" % trait\n",
    "    col5=\"%s.oref\" % trait\n",
    "   \n",
    "    df1=df1.rename(columns={0:\"Species\",1:\"Code\",2:col1,3:col2,4:col3,5:col4,6:col5})\n",
    "    df1[trait]=df1.apply(lambda row : fvx.summarise_values(row[col2],row[col3]), axis = 1)\n",
    "    if \"df\" in globals():\n",
    "        df = pd.merge(df, df1, on = [\"Species\",\"Code\"], how = \"outer\").sort_values(by=\"Species\",ascending=[1])\n",
    "    else:\n",
    "        df = df1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c959ee1e-188b-4820-a50a-de53cc65077f",
   "metadata": {},
   "source": [
    "Our dataframe `df` has multiple columns for each trait:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7b1449f1-83ed-44fd-89e9-c9f75657049b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Code</th>\n",
       "      <th>surv1.txn</th>\n",
       "      <th>surv1.v</th>\n",
       "      <th>surv1.w</th>\n",
       "      <th>surv1.mref</th>\n",
       "      <th>surv1.oref</th>\n",
       "      <th>surv1</th>\n",
       "      <th>surv4.txn</th>\n",
       "      <th>surv4.v</th>\n",
       "      <th>...</th>\n",
       "      <th>germ1.w</th>\n",
       "      <th>germ1.mref</th>\n",
       "      <th>germ1.oref</th>\n",
       "      <th>germ1</th>\n",
       "      <th>germ8.txn</th>\n",
       "      <th>germ8.v</th>\n",
       "      <th>germ8.w</th>\n",
       "      <th>germ8.mref</th>\n",
       "      <th>germ8.oref</th>\n",
       "      <th>germ8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abelmoschus moschatus subsp. moschatus</td>\n",
       "      <td>9878</td>\n",
       "      <td>[Abelmoschus moschatus]</td>\n",
       "      <td>[All]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[austraits-3.0.2]</td>\n",
       "      <td>[Clarke Lawes Murphy Russell-Smith Nano Bradst...</td>\n",
       "      <td>All</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abildgaardia ovata</td>\n",
       "      <td>8856</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Abildgaardia vaginata</td>\n",
       "      <td>9186</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Abrophyllum ornans</td>\n",
       "      <td>3220</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Abrotanella nivigena</td>\n",
       "      <td>1246</td>\n",
       "      <td>[Abrotanella nivigena]</td>\n",
       "      <td>[All]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[austraits-3.0.2]</td>\n",
       "      <td>[White Sinclair Frood 2020]</td>\n",
       "      <td>All</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6679</th>\n",
       "      <td>Zygophyllum iodocarpum</td>\n",
       "      <td>6357</td>\n",
       "      <td>[Roepera iodocarpa]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[austraits-3.0.2]</td>\n",
       "      <td>[White Sinclair Frood 2020]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6680</th>\n",
       "      <td>Zygophyllum ovatum</td>\n",
       "      <td>6358</td>\n",
       "      <td>[Roepera ovata]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[austraits-3.0.2]</td>\n",
       "      <td>[White Sinclair Frood 2020]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6681</th>\n",
       "      <td>Zygophyllum prismatothecum</td>\n",
       "      <td>6359</td>\n",
       "      <td>[Roepera prismatotheca]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[austraits-3.0.2]</td>\n",
       "      <td>[White Sinclair Frood 2020]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6682</th>\n",
       "      <td>Zygophyllum simile</td>\n",
       "      <td>10095</td>\n",
       "      <td>[Roepera similis]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[austraits-3.0.2]</td>\n",
       "      <td>[White Sinclair Frood 2020]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683</th>\n",
       "      <td>Zygophyllum spp.</td>\n",
       "      <td>ZYGO</td>\n",
       "      <td>[Zygophyllum sp.]</td>\n",
       "      <td>[None]</td>\n",
       "      <td>[1]</td>\n",
       "      <td>[austraits-3.0.2]</td>\n",
       "      <td>[White Sinclair Frood 2020]</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6686 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     Species   Code                surv1.txn   \n",
       "0     Abelmoschus moschatus subsp. moschatus   9878  [Abelmoschus moschatus]  \\\n",
       "1                         Abildgaardia ovata   8856                      NaN   \n",
       "2                      Abildgaardia vaginata   9186                      NaN   \n",
       "3                         Abrophyllum ornans   3220                      NaN   \n",
       "4                       Abrotanella nivigena   1246   [Abrotanella nivigena]   \n",
       "...                                      ...    ...                      ...   \n",
       "6679                  Zygophyllum iodocarpum   6357      [Roepera iodocarpa]   \n",
       "6680                      Zygophyllum ovatum   6358          [Roepera ovata]   \n",
       "6681              Zygophyllum prismatothecum   6359  [Roepera prismatotheca]   \n",
       "6682                      Zygophyllum simile  10095        [Roepera similis]   \n",
       "6683                        Zygophyllum spp.   ZYGO        [Zygophyllum sp.]   \n",
       "\n",
       "     surv1.v surv1.w         surv1.mref   \n",
       "0      [All]     [1]  [austraits-3.0.2]  \\\n",
       "1        NaN     NaN                NaN   \n",
       "2        NaN     NaN                NaN   \n",
       "3        NaN     NaN                NaN   \n",
       "4      [All]     [1]  [austraits-3.0.2]   \n",
       "...      ...     ...                ...   \n",
       "6679  [None]     [1]  [austraits-3.0.2]   \n",
       "6680  [None]     [1]  [austraits-3.0.2]   \n",
       "6681  [None]     [1]  [austraits-3.0.2]   \n",
       "6682  [None]     [1]  [austraits-3.0.2]   \n",
       "6683  [None]     [1]  [austraits-3.0.2]   \n",
       "\n",
       "                                             surv1.oref surv1 surv4.txn   \n",
       "0     [Clarke Lawes Murphy Russell-Smith Nano Bradst...   All       NaN  \\\n",
       "1                                                   NaN   NaN       NaN   \n",
       "2                                                   NaN   NaN       NaN   \n",
       "3                                                   NaN   NaN       NaN   \n",
       "4                           [White Sinclair Frood 2020]   All       NaN   \n",
       "...                                                 ...   ...       ...   \n",
       "6679                        [White Sinclair Frood 2020]  None       NaN   \n",
       "6680                        [White Sinclair Frood 2020]  None       NaN   \n",
       "6681                        [White Sinclair Frood 2020]  None       NaN   \n",
       "6682                        [White Sinclair Frood 2020]  None       NaN   \n",
       "6683                        [White Sinclair Frood 2020]  None       NaN   \n",
       "\n",
       "     surv4.v  ... germ1.w germ1.mref germ1.oref germ1 germ8.txn germ8.v   \n",
       "0        NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN  \\\n",
       "1        NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "2        NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "3        NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "4        NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "...      ...  ...     ...        ...        ...   ...       ...     ...   \n",
       "6679     NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "6680     NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "6681     NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "6682     NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "6683     NaN  ...     NaN        NaN        NaN   NaN       NaN     NaN   \n",
       "\n",
       "     germ8.w germ8.mref germ8.oref germ8  \n",
       "0        NaN        NaN        NaN   NaN  \n",
       "1        NaN        NaN        NaN   NaN  \n",
       "2        NaN        NaN        NaN   NaN  \n",
       "3        NaN        NaN        NaN   NaN  \n",
       "4        NaN        NaN        NaN   NaN  \n",
       "...      ...        ...        ...   ...  \n",
       "6679     NaN        NaN        NaN   NaN  \n",
       "6680     NaN        NaN        NaN   NaN  \n",
       "6681     NaN        NaN        NaN   NaN  \n",
       "6682     NaN        NaN        NaN   NaN  \n",
       "6683     NaN        NaN        NaN   NaN  \n",
       "\n",
       "[6686 rows x 44 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4953710f-5dbf-48cc-9fae-a309b68f4daa",
   "metadata": {},
   "source": [
    "The query string is different for numerical traits, so we update it and run another loop, this time with the numerical trait codes:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b18ffb81-dd0b-466f-bf2c-5c1ce75c2bc3",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "qry= \"\"\"\n",
    "SELECT \"currentScientificName\" as spp, \"currentScientificNameCode\" as sppcode,\n",
    "    array_agg(species) as nspp,\n",
    "    array_agg(best) as best,array_agg(lower) as lower,array_agg(upper) as upper,array_agg(weight) as w,\n",
    "    array_agg(main_source) as refs,\n",
    "    array_accum(original_sources) as orefs\n",
    "FROM litrev.{} \n",
    "LEFT JOIN species.caps\n",
    "ON species_code=\"speciesCode_Synonym\"\n",
    "WHERE \"currentScientificName\" is not NULL AND main_source is not NULL AND weight>0\n",
    "GROUP BY spp,sppcode;\n",
    "\"\"\"\n",
    "\n",
    "# test using only ecualipts\n",
    "# WHERE species ilike '%euca%' and \"currentScientificName\" is not NULL AND weight>0\n",
    "\n",
    "for trait in ['repr3','repr3a','repr4',]:\n",
    "    res = dbquery(qry.format(trait), dbparams)\n",
    "    if len(res)>0:\n",
    "        df1 = pd.DataFrame(res)\n",
    "        col1=\"%s.txn\" % trait\n",
    "        col2=\"%s.best\" % trait\n",
    "        col3=\"%s.lower\" % trait\n",
    "        col4=\"%s.upper\" % trait\n",
    "        col5=\"%s.w\" % trait\n",
    "        col6=\"%s.mref\" % trait\n",
    "        col7=\"%s.oref\" % trait\n",
    "        df1=df1.rename(columns={0:\"Species\",1:\"Code\",2:col1,3:col2,4:col3,5:col4,6:col5,7:col6,8:col7})\n",
    "        df1[trait]=df1.apply(lambda row : fvx.summarise_triplet(row[col2],row[col3],row[col4],row[col5]), axis = 1)\n",
    "        df = pd.merge(df, df1, on = [\"Species\",\"Code\"], how = \"outer\").sort_values(by=\"Species\",ascending=[1])\n",
    "   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2e38f41-6928-4310-a08b-a5e019848738",
   "metadata": {},
   "source": [
    "We can now apply these functions to summary data from multple columns into a single value:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "4a44542b-5e82-4131-a166-84ad767abbec",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['orig_species']=df.apply(lambda row : fvx.unique_taxa(row,'txn'), axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b4c75073-2d9a-4fa1-b062-05fbd524db8f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Code</th>\n",
       "      <th>surv1.txn</th>\n",
       "      <th>surv1.v</th>\n",
       "      <th>surv1.w</th>\n",
       "      <th>surv1.mref</th>\n",
       "      <th>surv1.oref</th>\n",
       "      <th>surv1</th>\n",
       "      <th>surv4.txn</th>\n",
       "      <th>surv4.v</th>\n",
       "      <th>...</th>\n",
       "      <th>repr3a</th>\n",
       "      <th>repr4.txn</th>\n",
       "      <th>repr4.best</th>\n",
       "      <th>repr4.lower</th>\n",
       "      <th>repr4.upper</th>\n",
       "      <th>repr4.w</th>\n",
       "      <th>repr4.mref</th>\n",
       "      <th>repr4.oref</th>\n",
       "      <th>repr4</th>\n",
       "      <th>orig_species</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1494</th>\n",
       "      <td>Chenopodium desertorum</td>\n",
       "      <td>2091</td>\n",
       "      <td>[Chenopodium desertorum, Chenopodium desertoru...</td>\n",
       "      <td>[None, None, None, None, None, None]</td>\n",
       "      <td>[1, 1, 10, 1, 1, 1]</td>\n",
       "      <td>[austraits-3.0.2, austraits-3.0.2, NSWFFRDv2.1...</td>\n",
       "      <td>[Vesk Leishman Westoby 2004, White Sinclair Fr...</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Chenopodium desertorum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 69 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     Species  Code   \n",
       "1494  Chenopodium desertorum  2091  \\\n",
       "\n",
       "                                              surv1.txn   \n",
       "1494  [Chenopodium desertorum, Chenopodium desertoru...  \\\n",
       "\n",
       "                                   surv1.v              surv1.w   \n",
       "1494  [None, None, None, None, None, None]  [1, 1, 10, 1, 1, 1]  \\\n",
       "\n",
       "                                             surv1.mref   \n",
       "1494  [austraits-3.0.2, austraits-3.0.2, NSWFFRDv2.1...  \\\n",
       "\n",
       "                                             surv1.oref surv1 surv4.txn   \n",
       "1494  [Vesk Leishman Westoby 2004, White Sinclair Fr...  None       NaN  \\\n",
       "\n",
       "     surv4.v  ... repr3a repr4.txn repr4.best repr4.lower repr4.upper repr4.w   \n",
       "1494     NaN  ...    NaN       NaN        NaN         NaN         NaN     NaN  \\\n",
       "\n",
       "     repr4.mref repr4.oref repr4            orig_species  \n",
       "1494        NaN        NaN   NaN  Chenopodium desertorum  \n",
       "\n",
       "[1 rows x 69 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[1494:1495]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c0da2808-c536-420e-9c82-d24831e9dea5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df['main_refs']=df.apply(lambda row : fvx.unique_taxa(row,'mref'), axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3ed46670-4cf4-4906-9737-ee6fa42f68e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig_species</th>\n",
       "      <th>main_refs</th>\n",
       "      <th>orig_refs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Abelmoschus moschatus</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>Clarke Lawes Murphy Russell-Smith Nano Bradsto...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Abildgaardia ovata</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>Metcalfe 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Abildgaardia vaginata</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>Metcalfe 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Abrophyllum ornans</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>Hughes Rice 1992; Cooper Cooper 2013</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Abrotanella nivigena</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>White Sinclair Frood 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6681</th>\n",
       "      <td>Roepera iodocarpa</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>White Sinclair Frood 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6682</th>\n",
       "      <td>Roepera ovata</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>White Sinclair Frood 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683</th>\n",
       "      <td>Roepera prismatotheca</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>White Sinclair Frood 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6684</th>\n",
       "      <td>Roepera similis</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>White Sinclair Frood 2020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6685</th>\n",
       "      <td>Zygophyllum sp.</td>\n",
       "      <td>austraits-3.0.2</td>\n",
       "      <td>White Sinclair Frood 2020</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6686 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               orig_species        main_refs   \n",
       "0     Abelmoschus moschatus  austraits-3.0.2  \\\n",
       "1        Abildgaardia ovata  austraits-3.0.2   \n",
       "2     Abildgaardia vaginata  austraits-3.0.2   \n",
       "3        Abrophyllum ornans  austraits-3.0.2   \n",
       "4      Abrotanella nivigena  austraits-3.0.2   \n",
       "...                     ...              ...   \n",
       "6681      Roepera iodocarpa  austraits-3.0.2   \n",
       "6682          Roepera ovata  austraits-3.0.2   \n",
       "6683  Roepera prismatotheca  austraits-3.0.2   \n",
       "6684        Roepera similis  austraits-3.0.2   \n",
       "6685        Zygophyllum sp.  austraits-3.0.2   \n",
       "\n",
       "                                              orig_refs  \n",
       "0     Clarke Lawes Murphy Russell-Smith Nano Bradsto...  \n",
       "1                                         Metcalfe 2020  \n",
       "2                                         Metcalfe 2020  \n",
       "3                  Hughes Rice 1992; Cooper Cooper 2013  \n",
       "4                             White Sinclair Frood 2020  \n",
       "...                                                 ...  \n",
       "6681                          White Sinclair Frood 2020  \n",
       "6682                          White Sinclair Frood 2020  \n",
       "6683                          White Sinclair Frood 2020  \n",
       "6684                          White Sinclair Frood 2020  \n",
       "6685                          White Sinclair Frood 2020  \n",
       "\n",
       "[6686 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['orig_refs']=df.apply(lambda row : fvx.unique_taxa(row,'oref'), axis = 1)\n",
    "df[['orig_species','main_refs','orig_refs']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebe871ff-4f45-496a-a27a-1718df8d793e",
   "metadata": {},
   "source": [
    "And now we can extract a list of valid references from this data frame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "40b11453-bf21-44c1-8f0e-e2496cbdd131",
   "metadata": {},
   "outputs": [],
   "source": [
    "refs=df.apply(lambda row : fvx.extract_refs(row,'mref'), axis = 1)\n",
    "valid_refs=list()\n",
    "for x in refs:\n",
    "    if type(x)==list:\n",
    "        valid_refs=valid_refs+x\n",
    "        \n",
    "refs=df.apply(lambda row : fvx.extract_refs(row,'oref'), axis = 1)\n",
    "for x in refs:\n",
    "    if type(x)==list:\n",
    "        valid_refs=valid_refs+x\n",
    "   \n",
    "valid_refs=tuple(set(valid_refs))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42ffe44e-1950-4e5b-8e99-1462587b205c",
   "metadata": {},
   "source": [
    "And now query the database to include only references in that list:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3d2f0213-98ac-491d-a0f5-7be6ca033bcc",
   "metadata": {},
   "outputs": [],
   "source": [
    "qrystr = \"SELECT ref_code,ref_cite FROM litrev.ref_list WHERE ref_code IN %s ORDER BY ref_code\" % (valid_refs,)\n",
    "ref_info = dbquery(qrystr, dbparams)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "240d0f41-29a3-48dc-a85c-68c8be9a9e58",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Create workbook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "36492b0e-fd7f-4c1f-aa47-e29a8a55fff6",
   "metadata": {},
   "outputs": [],
   "source": [
    "wb = Workbook()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1edfb5de-a5f6-4cf5-aeca-20aadd1a99ee",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Styles\n",
    "Define styles to be used across the workbook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "9ec51a4e-1bce-4990-8a4b-c1a95ad172b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cent_align=Alignment(horizontal='center', vertical='center', wrap_text=False)\n",
    "wrap_align=Alignment(horizontal='left', vertical='top', wrap_text=True)\n",
    "\n",
    "fontSmall = Font(size = \"9\")\n",
    "\n",
    "\n",
    "sheet_colors = {\"intro\": \"1072BA\" , \"summary\": \"5AFF5A\", \"default\":\"505050\", \"addentry\": \"20CA82\"}\n",
    "\n",
    "table_style={\"Instructions\":TableStyleInfo(name=\"TableStyleMedium9\", showFirstColumn=True, showLastColumn=False, \n",
    "                                           showRowStripes=True, showColumnStripes=False),\n",
    "             \"Contributor\": TableStyleInfo(name=\"TableStyleMedium18\", showFirstColumn=True,\n",
    "                       showLastColumn=False, showRowStripes=False, showColumnStripes=False),\n",
    "             \"Lists\": TableStyleInfo(name=\"TableStyleMedium14\", showFirstColumn=True,\n",
    "                       showLastColumn=False, showRowStripes=False, showColumnStripes=False),\n",
    "             \"Info\":  TableStyleInfo(name=\"TableStyleMedium14\", showFirstColumn=True,\n",
    "                       showLastColumn=False, showRowStripes=False, showColumnStripes=False),\n",
    "             \"Vocabularies\": TableStyleInfo(name=\"TableStyleMedium14\", showFirstColumn=True,\n",
    "                       showLastColumn=False, showRowStripes=False, showColumnStripes=False),\n",
    "             \"Entry\": TableStyleInfo(name=\"TableStyleMedium18\", showFirstColumn=False,\n",
    "                       showLastColumn=False, showRowStripes=False, showColumnStripes=False)\n",
    "\n",
    "             }\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9d8fd4f-553c-4cf8-957e-51980fdee234",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create worksheets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e7ff02df-a31a-4c68-93ec-9edb8980b453",
   "metadata": {},
   "outputs": [],
   "source": [
    "wsheets = (\n",
    "    {\"title\": \"About\", \"colWidths\":[(\"A\",90),(\"B\",40)], \"tabColor\":\"intro\",\"active\":True},\n",
    "    {\"title\": \"Summary\", \"colWidths\":[(\"A\",70),(\"B\",10),((\"C\",\"D\",\"E\",\"F\",\"G\",\"H\",\"I\",\"J\",\"K\"),30),((\"L\",\"M\",\"N\",),25)], \"tabColor\":\"summary\"},\n",
    "    {\"title\": \"References\", \"colWidths\":[(\"A\",25),(\"B\",80)], \"tabColor\":\"addentry\"},\n",
    "    {\"title\": \"Trait description\", \"colWidths\":[(\"A\",12),(\"B\",30),(\"C\",70)], \"tabColor\":\"default\"}\n",
    "    )\n",
    "for item in wsheets:\n",
    "    if \"active\" in item.keys():\n",
    "        ws = wb.active\n",
    "        ws.title = item['title']\n",
    "    else:\n",
    "        ws = wb.create_sheet(item['title'])\n",
    "    for k in item['colWidths']:\n",
    "        for j in k[0]:\n",
    "            ws.column_dimensions[j].width = k[1]\n",
    "    ws.sheet_properties.tabColor = sheet_colors[item[\"tabColor\"]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bff0687-aa8f-41f7-87a6-57ec57cbe9f1",
   "metadata": {
    "tags": []
   },
   "source": [
    "### `About` worksheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "007df909-b552-4606-8151-466fe16aba9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ws = wb[\"About\"]\n",
    "\n",
    "info = (\"Fire Ecology Traits for Plants\",\n",
    "        \"Version 1.00 (April 2022)\",\n",
    "        \"This data export reflects the status of the database on the %s\" % datetime.date.today().strftime('%d %b %Y'),\n",
    "        \"Developed by  José R. Ferrer-Paris and David Keith\",\n",
    "        \"Centre for Ecosystem Science / University of New South Wales\",\n",
    "        \"Please cite this work as:\",\n",
    "        \"Ferrer-Paris, J. R. and Keith, D. A. (2024) Fire Ecology Traits for Plants: Database export. figshare. DOI: 10.6084/m9.figshare.24125088\", \n",
    "        #\"DISCLAIMER:\",\n",
    "        #\"DATA IS NOT READY FOR FINAL USE OR CRITICAL APPLICATIONS AND YOU SHOULD NOT DISTRIBUTE THIS DATA.\"\n",
    "        )\n",
    "\n",
    "k = 1\n",
    "for row in info:\n",
    "    ws.cell(k,1,value=row)\n",
    "    ws.cell(k,1).alignment=wrap_align\n",
    "    k=k+1\n",
    "    \n",
    "ws.cell(1,1).style='Title'\n",
    "ws.cell(5,1).hyperlink='https://www.unsw.edu.au/research/ecosystem'\n",
    "ws.cell(5,1).style='Hyperlink'\n",
    "\n",
    "# Disclaimer\n",
    "ws.cell(8,1).font=Font(color=\"FF0000\", bold=True,italic=False) \n",
    "ws.cell(9,1).font=Font(color=\"FF0000\", italic=True) \n",
    "\n",
    "\n",
    "supporters = ({'institution':\"University of New South Wales\",'url':\"https://www.unsw.edu.au/\"},\n",
    "              {'institution':\"NSW Bushfire Research Hub\",'url':\"https://www.bushfirehub.org/\"},\n",
    "              {'institution':\"NESP Threatened Species Recovery Hub\",'url':\"https://www.nespthreatenedspecies.edu.au/\"},\n",
    "              {'institution':\"NSW Department of Planning & Environment\",'url':\"https://www.planning.nsw.gov.au/\"})\n",
    "\n",
    "k=k+2\n",
    "ws.cell(k-1,1,value=\"This work has been supported by:\")\n",
    "for item in supporters:\n",
    "    cell=ws.cell(k,1)\n",
    "    cell.value=item['institution']\n",
    "    cell.hyperlink=item['url']\n",
    "    cell.style = \"Hyperlink\"\n",
    "    k=k+1\n",
    "\n",
    "k=k+2\n",
    "description = (\n",
    "              \"Taxonomic nomenclature following BioNET (data export from February 2022)\",\n",
    "              \"Data in the report is summarised based on BioNET fields 'currentScientificName' and 'currentScientificNameCode'\",\n",
    "              \"For general description of the traits, please refer to the 'Trait description' sheet\",\n",
    "              \"Vocabularies for categorical traits are available in the 'Vocabularies' sheet\",\n",
    "              \"For categorical traits the values in the 'Summary' sheet show the different values reported in the literature records separated by slashes.\",\n",
    "               \"If more than one category has been reported, the values are ordered from higher to lower 'weight', categories receiving less than 10% weight are in round brackets, categories with less than 5% in square brackets\",\n",
    "              \"The default weight is calculated by multiplying the number of times a value is reported (nr. of records) with the weight given to each record (default to 1), and divided by the weight of all records for a given species.\",\n",
    "              \"Default weights  overridden by expert advice to the administrator will be marked, with justification given in the Notes column of the output.\",\n",
    "              \"An asterisk (*) in a trait cell indicates a potential data entry error or uncertainty in the assignment of a trait category or value.\",\n",
    "              \"'Import/Entry sources' refer to references that were imported directly using automated scripts or manual entry. These include: 1) Primary observations of traits from published research or reports; and 2) Compilations of data (e.g. databases, spreadsheets, published reviews) that include two or more sources of primary observations.\",\n",
    "              \"'Indirect sources' refer to references that were cited in Import/Entry sources, where the latter are compilations of multiple primary sources (see Import/Entry sources). Information from indirect sources may have been modified when it was incorporated into those compilations. The original source of primary trait observations has not yet been verified prior to import into this database. When the primary source is reviewed and the trait values are verified, these records will be attributed to the primary source as 'Import/Entry sources'.\",\n",
    "              \"Some sheets are protected to avoid accidental changes, but they are not password protected. If you need to filter and reorder entries in the table, please unprotect the sheet first.\",\n",
    "              )\n",
    "\n",
    "for row in description:\n",
    "    ws.cell(k,1,value=row)\n",
    "    ws.cell(k,1).alignment=wrap_align\n",
    "    k=k+1\n",
    "    \n",
    "ws.protection.sheet = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa22fc81-1c98-44e9-8482-1738cbc72365",
   "metadata": {
    "tags": []
   },
   "source": [
    "### `Trait description` worksheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a37de598-e190-4e0e-b3a1-cc9d4f7d86fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "ws = wb[\"Trait description\"]\n",
    "\n",
    "k=1\n",
    "description = (\"The following table gives a general description of the traits used in the 'Summary' sheet\",\n",
    "               \"This sheet is protected to avoid accidental changes, but it is not password protected. If you need to filter and reorder entries in the table, please unprotect the sheet first.\",\n",
    "              \"Vocabularies for categorical traits are available in the 'Vocabularies' sheet\",\"\",\"\")\n",
    "\n",
    "for row in description:\n",
    "    ws.cell(k,3,value=row)\n",
    "    ws.cell(k,3).alignment=wrap_align\n",
    "    k=k+1\n",
    "    \n",
    "\n",
    "ws.append([\"Trait Code\", \"Trait Name\", \"Description\", \"Type\", \"Life stage\", \"Life history process\", \"Data migration\"])\n",
    "\n",
    "for row in trait_info:\n",
    "    ws.append(row)\n",
    "    \n",
    "#ws.max_row\n",
    "for j in range(k,ws.max_row+1):\n",
    "    ws.cell(j,3).alignment=wrap_align\n",
    "    \n",
    "tab = Table(displayName=\"TraitInformation\", ref=\"A{}:G{}\".format(k,ws.max_row))\n",
    "\n",
    "tab.tableStyleInfo = table_style[\"Info\"]\n",
    "ws.add_table(tab)\n",
    "ws.protection.sheet = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cea2517e-5920-462e-a1b1-e79fe282f2f4",
   "metadata": {
    "tags": []
   },
   "source": [
    "### `Summary` worksheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e44102c5-2092-4375-8bf0-5a1078e7ce88",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "ws = wb[\"Summary\"]\n",
    "ws.append(['Species','Code','surv1','surv4','germ1','germ8','rect2','repr2','repr3','repr3a','disp1','Original Species name(s) used','Import/Entry sources','Indirect sources'])\n",
    "rows = dataframe_to_rows(df[['Species','Code','surv1','surv4','germ1','germ8','rect2','repr2','repr3','repr3a','disp1','orig_species','main_refs','orig_refs']],index=False, header=False)\n",
    "\n",
    "\n",
    "for r_idx, row in enumerate(rows, 2):\n",
    "    for c_idx, value in enumerate(row, 1):\n",
    "        ws.cell(row=r_idx, column=c_idx, value=value)\n",
    "    \n",
    "    for k in (12,13,14):\n",
    "        ws.cell(r_idx,k).alignment=wrap_align\n",
    "        ws.cell(r_idx,k).font = fontSmall\n",
    "\n",
    "    \n",
    "tab = Table(displayName=\"Summary\", ref=\"A1:{}{}\".format(get_column_letter(c_idx),r_idx))\n",
    "tab.tableStyleInfo = table_style[\"Lists\"]\n",
    "ws.add_table(tab)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9a2b9e7-5f46-44bd-9a51-fc37ddb99a48",
   "metadata": {},
   "source": [
    "### `References` worksheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "92d7d546-9991-49ae-9124-db09c0c17b26",
   "metadata": {},
   "outputs": [],
   "source": [
    "ws = wb[\"References\"]\n",
    "\n",
    "k=1\n",
    "description = (\"The following table includes bibliographical information for the sources referenced in the 'Summary' sheet\",\n",
    "               \"This sheet is protected to avoid accidental changes, but it is not password protected. If you need to filter and reorder entries in the table, please unprotect the sheet first.\",\n",
    "              \"\",\"\")\n",
    "\n",
    "for row in description:\n",
    "    ws.cell(k,2,value=row)\n",
    "    ws.cell(k,2).alignment=wrap_align\n",
    "    k=k+1\n",
    "    \n",
    "\n",
    "ws.append([\"Reference code\", \"Reference information\"])\n",
    "\n",
    "for row in ref_info:\n",
    "    ws.append(row)\n",
    "    \n",
    "#ws.max_row\n",
    "for j in range(k+1,ws.max_row+1):\n",
    "    ws.cell(j,2).alignment=wrap_align\n",
    "    ws.cell(j,2).font = fontSmall\n",
    "    \n",
    "tab = Table(displayName=\"ReferenceInformation\", ref=\"A{}:B{}\".format(k,ws.max_row))\n",
    "\n",
    "tab.tableStyleInfo = table_style[\"Lists\"]\n",
    "ws.add_table(tab)\n",
    "ws.protection.sheet = True"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98e9832f-eef4-4c51-bf8e-6de0224f6699",
   "metadata": {},
   "source": [
    "### Save workbook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "499c421f-bba6-45ab-a147-bc43a4d741a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "wb.save(inputdir / \"fireveg-trait-report-model.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e44e4e89-6e05-4fdc-9205-f2364c637c73",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
